Byron wallet restoration from xprv endpoint impl

paweljakubas opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->
#1436 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have worked on separating attachPrivateKey into two groups using different encryption
- [x] I have added  implementation of `generateKeyFromMasterKey` in `AddressDerivation.Byron`
- [x] I have added implementation of `postRandomWalletFromXPrv`
- [x] I have added corresponding unit tests
- [ ] I have added illustrative integration tests  (in next PR)


# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

paweljakubas committed:
extract attachPrivateKeyFromPwd

paweljakubas committed:
introduce Scrypt hash attach private key function in Cardano.Wallet

paweljakubas committed:
implement postRandomWalletFromXPrv

paweljakubas committed:
add implementation of generateKeyFromMasterKey in AddressDerivation.Byron

correct hdPassword

paweljakubas committed:
add unit tests

KtorZ committed:
Augment passphrase metadata info with encryption scheme

This to allow distinguish legacy passphrase hashes imported from Daedalus that are
encrypted through Scrypt and the new passphrases that we now encrypt with PBKDF2.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Byron.hs:
```diff
@@ -266,6 +269,21 @@ hdPassphrase masterKey = Passphrase $
+    { getKey = masterKey
+    , derivationPath
+    , payloadPassphrase = hdPassphrase (toXPub masterKey)
+    }
```
These functions aren't generating any keys though :grimacing: 
I don't think they are necessary to be honest, the calling code could simply be calling `ByronKey` directly, or, maybe we can simply rename the function to `mkByronKey :: XPrv -> ByronKey 'RootK XPrv`

paweljakubas replied:
yes, fair remarks. `mkByronKey` is better naming

paweljakubas replied:
done in https://github.com/input-output-hk/cardano-wallet/pull/1454/commits/af27d7004db13e835c679bfa6d78a9692ed7b28d

KtorZ committed:
Actually implement 'checkPassphrase' and 'encryptPassphrase' for Scrypt

This still misses some test vectors to compare with the legacy implementation. I'll
produce that soon using both Daedalus + cardano-sl code.

paweljakubas committed:
add integration test - part 1

paweljakubas committed:
rename generate to mkByron

KtorZ committed:
rework the checkPassphrase / encryptPassphrase phantom types

Wallet using a passphrase that is encrypted with the legacy scheme needs to go through some special
transformation before being encrypted. In order to present programmer mistakes here, we leverage a bit
the phantom types for 'Passphrase' such that passphrases obtained from the API are marked as 'raw' whereas,
every function from the 'AddressDerivation' module requires a passphrase to be marked as 'encryption'.
Such passphrases can be obtained via 'preparePassphrase' which does nothing but a simple type coercion,
for legacy passphrases it'll do some preliminary transformation.

KtorZ committed:
add golden test about legacy passphrase encryption

The golden string was obtained from cardano-sl, using a raw string that was blake2b_256-encoded by
Daedalus.

KtorZ commented on lib/core-integration/src/Test/Integration/Scenario/API/ByronWallets.hs:
```diff
@@ -399,6 +407,21 @@ spec = do
+
+        let (Right seed) = fromMnemonic @'[12] m
+        let _rootXPrv =
+                generateKeyFromSeed seed (Passphrase $ BA.convert $ T.encodeUtf8 passw)
```
@paweljakubas may you clarify what this call is for :thinking: ? Or is there actually an API call missing from here?

paweljakubas replied:
yeah, this is not finished integration tests. I was waiting for your branch to me merged to use stuff hashing password. As agreed we will do it in next PR

KtorZ replied:
Ah! Damn. I actually removed that from 'encryptPassphrase` because there's actually no reason we would try to encrypt a passphrase with Scrypt on the new codebase :s 

May I suggest to use this passphrase hash instead:

```
31347c387c317c574342652b796362417576356c2b4258676a344a314c
6343675375414c2f5653393661364e576a2b7550766655513d3d7c2f37
6738486c59723174734e394f6e4e753253302b6a65515a6b5437316b45
414941366a515867386539493d
```

It was generated using cardano-sl, it corresponds to the passphrase "patate".

KtorZ committed:
re-generate nix machinery

KtorZ committed:
apply HLintsuggestion about using newtype

KtorZ committed:
add missing update field for passphrase scheme in SQLite

KtorZ committed:
add new golden files for API
